.. _docs_gp_dnaseq:

.. spelling:: 
 
     haplotype
     Mpileup
     mpileup
     bcftools
     kilobases
     telomere
     mappability
     indel
     dna
     callsets
     intra
     nucleotid
     tdf

DNA Sequencing Pipeline
========================

DNA sequencing is the process of determining the sequence of nucleotides in a section of DNA. Next-generation sequencing (NGS), also known as high-throughput sequencing, allows for sequencing of DNA and RNA much more quickly and cheaply than the previously used Sanger sequencing, and as such revolutionized the study of genomics and molecular biology. The enormous amount of short reads generated by the new DNA sequencing technologies call for the development of fast and accurate read alignment programs. 

Over the past decade, `long-read`_, single-molecule DNA sequencing technologies have emerged as powerful players in genomics. With the ability to generate reads tens to thousands of kilobases in length with an accuracy approaching that of short-read sequencing technologies, these platforms have proven their ability to resolve some of the most challenging regions of the human genome, detect previously inaccessible structural variants and generate some of the first telomere-to-telomere assemblies of whole chromosomes.

Burrows-Wheeler Alignment tool, `BWA`_, is a new read alignment package that is based on backward search with Burrows–Wheeler Transform (BWT), to efficiently align short sequencing reads against a large reference sequence such as the human genome, allowing mismatches and gaps. BWA is ∼10–20 times `faster than`_ Mapping and Assembly with Quality, `MAQ`_, while achieving similar accuracy. In addition, BWA outputs alignment in the new standard SAM (Sequence Alignment/Map) format. Variant calling and other downstream analyses after the alignment can be achieved with the open source `SAMtools software package`_.

DNA Sequencing GenPipes pipeline has been implemented by optimizing the Genome Analysis Toolkit, `GATK`_ best practices standard operating protocols. This procedure entails trimming raw reads derived from whole-genome or exome data followed by alignment to a known reference, post-alignment refinements, and variant calling. Trimmed reads are aligned to a reference by the Burrows-Wheeler Aligner (BWA), `bwa-mem`_. Refinements of mismatches near insertions and deletions (indels) and base qualities are performed using GATK indels realignment and base recalibration to improve read quality after alignment. Processed reads are marked as fragment duplicates using Picard MarkDuplicates and single-nucleotide polymorphisms and small indels are identified using either GATK haplotype callers or `SAMtools`_ mpileup. 

The `Genome in a Bottle dataset`_ was used to select steps and parameters minimizing the false-positive rate and maximizing the true-positive variants to achieve a sensitivity of 99.7%, precision of 99.1%, and F1 score of 99.4%. Finally, additional annotations are incorporated using `dbNSFP`_ and / or Gemini and QC metrics are collected at various stages and visualized using `MultiQC`_. 

Latest release 3.2.0 of GenPipes supports a new DNA sequencing protocol option called 'sv' besides the mugqic, mpileup and light options. This is useful for structural variation detection.

`Structural Variation`_ (SVs) are the genetic variations in the structure of chromosome with different types of rearrangements. They comprise millions of nucleotides of heterogeneity within every genome, and are likely to make an important contribution to genetic diversity and disease susceptibility. In the genomics community, substantial efforts have been devoted to improving understanding of the roles of SVs in genome functions relating to diseases and researchers are working actively to develop effective algorithms to reliably identify various types of SVs such as deletions, insertions, duplications and inversions. GenPipes supports SV detection option in the DNA Sequencing Pipeline.

.. contents:: :local:

----

Introduction
------------

The standard MUGQIC DNA-Seq pipeline uses BWA to align reads to the reference genome. Treatment and filtering of mapped reads approaches as INDEL realignment, mark duplicate reads, recalibration and sort are executed using Picard and GATK. Samtools MPILEUP and bcftools are used to produce the standard SNP and indels variants file (VCF). Additional SVN annotations mostly applicable to human samples include mappability flags, dbSNP annotation and extra information about SVN by using published databases.  The SNPeff tool is used to annotate variants using an integrated database of functional predictions from multiple algorithms (SIFT, Polyphen2, LRT and MutationTaster, PhyloP and GERP++, etc.) and to calculate the effects they produce on known genes. Refer to the list of `SnpEff effects`_.

A summary html report is automatically generated by the pipeline. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results.  Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters that have been passed to the pipeline main script.  

The DNA sequencing pipeline supports two trimmers: `Trimmomatic`_ and `Skewer <https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-182>`_.

GenPipes DNA sequencing pipeline offers four different protocol options:

#. The default protocol based on the GATK variant caller, haplotype caller, (-t mugqic)
#. Another based on the mpileup/bcftools caller (-t mpileup). 
#. Light option (-t light)
#. Structural Variation Detection, sv option (-t sv)

See :ref:`More Information on DNA Sequencing` section below for details. 

----

Version
-------

|genpipes_version|

For the latest implementation and usage details refer to DNA Sequencing implementation `README file <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/dnaseq/README.md>`_ file.

----

Usage
-----

::

  dnaseq.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
            [-o OUTPUT_DIR] [-j {pbs,batch,daemon,slurm}] [-f]
            [--no-json] [--report] [--clean]
            [-l {debug,info,warning,error,critical}] [--sanity-check]
            [--container {wrapper, singularity} <IMAGE PATH>
            [--genpipes_file GENPIPES_FILE]
            [-t {mugqic,mpileup,light,sv}] [-r READSETS] [-v]

**Optional Arguments**

.. include:: opt_dnaseq.inc
.. include:: /common/gp_readset_opt.inc
.. include:: /common/gp_common_opt.inc

----
 
Example Run
-----------

Use the following commands to execute MUGQIC DNA sequencing pipeline:

.. include::  /user_guide/pipelines/example_runs/dnaseq_mugqic.inc

Use the following commands to execute the Mpileup DNA sequencing pipeline:

.. include::  /user_guide/pipelines/example_runs/dnaseq_mpileup.inc

.. include:: /user_guide/pipelines/notes/scriptfile_deprecation.inc

You can download the test dataset for this pipeline :ref:`here<docs_testdatasets>`.

---- 

Pipeline Schema
---------------

Figure below shows the schema of the DNA sequencing protocol - MUGQIC type. 

.. figure:: /img/pipelines/mmd/dnaseq.mugqic.mmd.png
   :align: center
   :alt: dnaseq1 schema
   :width: 100%
   :figwidth: 95%

   Figure: Schema of MUGQIC DNA Sequencing protocol

.. figure:: /img/pipelines/mmd/legend.mmd.png
   :align: center
   :alt: dada2 ampseq
   :width: 100%
   :figwidth: 75%

The following figure shows the pipeline schema for `Mpileup`_ type of DNA sequencing protocol. 

.. figure:: /img/pipelines/mmd/dnaseq.mpileup.mmd.png
   :align: center
   :alt: dnaseq2 schema 
   :width: 100%
   :figwidth: 95%

   Figure: Schema of Mpileup DNA Sequencing protocol

.. figure:: /img/pipelines/mmd/legend.mmd.png
   :align: center
   :alt: dada2 ampseq
   :width: 100%
   :figwidth: 75%


The following figure shows the pipeline schema for 'Light' type of DNA sequencing protocol. 

.. figure:: /img/pipelines/mmd/dnaseq.light.mmd.png
   :align: center
   :alt: dnaseq3 schema 
   :width: 100%
   :figwidth: 95%

   Figure: Schema of Light DNA Sequencing protocol

.. figure:: /img/pipelines/mmd/legend.mmd.png
   :align: center
   :alt: dada2 ampseq
   :width: 100%
   :figwidth: 75%

The following figure shows the pipeline schema for `Structural Variation`_ type of DNA sequencing protocol. 

.. figure:: /img/pipelines/mmd/dnaseq.sv.mmd.png
   :align: center
   :alt: dnaseq4 schema 
   :width: 100%
   :figwidth: 95%

   Figure: Schema of Structural Variations (SV) DNA Sequencing protocol

.. figure:: /img/pipelines/mmd/legend.mmd.png
   :align: center
   :alt: dada2 ampseq
   :width: 100%
   :figwidth: 75%

Refer to the Pipeline steps below, for details.

----

Pipeline Steps
--------------

The table below shows various steps that constitute the MUGQIC and Mpileup types of DNA Sequencing for genomic analysis pipelines.

+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
|    |  *MUGQIC DNA Sequencing Steps* |   *MPileup DNA Sequencing Steps*| *Light DNA Sequencing Steps*    | *Structural Variations Steps*  |
+====+================================+=================================+=================================+================================+
| 1. | |picard_sam_to_fastq|          | |picard_sam_to_fastq|           | |picard_sam_to_fastq|           | |picard_sam_to_fastq|          |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 2. | |skewer_trim|                  | |skewer_trim|                   | |skewer_trim|                   | |skewer_trim|                  |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 3. | |bwa_mem_sambamba_sort_sam|    | |bwa_mem_sambamba_sort_sam|     | |bwa_mem_sambamba_sort_sam|     | |bwa_mem_sambamba_sort_sam|    |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 4. | |sambamba_merge|               | |sambamba_merge|                | |sambamba_merge|                | |sambamba_merge|               |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 5. | |gatk_indel_realigner|         | |gatk_indel_realigner|          | |gatk_indel_realigner|          | |gatk_indel_realigner|         |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 6. | |sambamba_merge_realigned|     | |sambamba_merge_realigned|      | |sambamba_merge_realigned|      | |sambamba_merge_realigned|     |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 7. | |picard_mark_dup|              | |picard_mark_dup|               | |picard_mark_dup|               | |picard_mark_dup|              |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 8. | |recalib|                      | |recalib|                       | |recalib|                       | |recalib|                      |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 9. | |gatk_hp_c|                    | |rawmpileup|                    | |sym_link_final_bam|            | |gatk_hp_c|                    |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 10.| |merge_call_i_gvcf|            | |rawmp_cat|                     | |m_dna_picard|                  | |merge_call_i_gvcf|            |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 11.| |combine_gvcf|                 | |snp_n_indel_bcf|               | |m_dna_sample|                  | |m_dna_picard|                 |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 12.| |merge_call_c_gvcf|            | |merge_filter_bcf|              | |m_dna_sambamba|                | |delly_call_filter|            |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 13.| |variant_recalib|              | |mp_dcom_norm|                  | |m_dna_fastqc|                  | |delly_sv_annotation|          |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 14.| |hc_dec_norm|                  | |mp_flag_map|                   | |picard_calc_hs_m|              | |manta_sv_calls|               |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 15.| |hc_flag_map|                  | |mp_snp_id_ann|                 | |gatk_call_loci|                | |manta_sv_annotation|          |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 16.| |hc_snp_ann|                   | |mp_snp_eff|                    | |extract_com_snp|               | |lumpy_paired_sv|              |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 17.| |hc_snp_eff|                   | |mp_dbnsfp_ann|                 | |baf_plot|                      | |lumpy_sv_annotation|          |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 18.| |hc_dbnsfp_ann|                | |mp_gemini_ann|                 | |gatk_hp_c|                     | |wham_sv_call|                 |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 19.| |hc_gemini_ann|                | |mp_m_vcf_stat|                 | |merge_call_i_gvcf|             | |wham_sv_annotation|           |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 20.| |m_dna_picard|                 | |cram_output|                   | |combine_gvcf|                  | |cnvkit_batch|                 |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 21.| |m_dna_sample|                 | |m_dna_picard|                  | |merge_call_c_gvcf|             | |cnvkit_sv_annotation|         |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 22.| |m_dna_fastqc|                 | |m_dna_sample|                  | |variant_recalib|               | |run_breakseq2|                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 23.| |picard_calc_hs_m|             | |m_dna_fastqc|                  | |hc_dec_norm|                   | |ensemble_metasv|              |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 24.| |metrics|                      | |picard_calc_hs_m|              | |hc_flag_map|                   | |metasv_sv_annotation|         |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 25.| |gatk_call_loci|               | |gatk_call_loci|                | |hc_snp_ann|                    |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 26.| |extract_com_snp|              | |extract_com_snp|               | |hc_snp_eff|                    |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 27.| |baf_plot|                     | |baf_plot|                      | |hc_dbnsfp_ann|                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 28.| |run_multiqc|                  | |run_multiqc|                   | |hc_gemini_ann|                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 29.| |cram_output|                  | |sym_link_fastq|                | |run_multiqc|                   |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 30 | |sym_link_fastq|               | |sym_link_final_bam|            | |cram_output|                   |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 31.| |sym_link_final_bam|           |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 32.| |m_ngscheckmate|               |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 33.| |m_verify_bam_id|              |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 34.| |m_vcftools_missing_i|         |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 35.| |m_vcftools_depth_i|           |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 36.| |m_gatk_sample_fp|             |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+
| 37.| |m_gatk_cluster_fp|            |                                 |                                 |                                |
+----+--------------------------------+---------------------------------+---------------------------------+--------------------------------+

----

.. include:: steps_dnaseq.inc

----

.. _More Information on DNA Sequencing:

More information
-----------------

For the latest implementation and usage details refer to DNA Sequencing implementation `README.md <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/dnaseq/README.md>`_ file.

* `Three-stage quality control strategies for DNA sequencing <https://academic.oup.com/bib/article/15/6/879/180439>`_

* `NGS Mapping, errors and quality control <http://bioinformatics.org.au/ws14/wp-content/uploads/ws14/sites/5/2014/07/Felicity-Newell_presentation.pdf>`_

* `dbNSFP: a lightweight database of human nonsynonymous SNPs and their functional predictions <https://www.ncbi.nlm.nih.gov/pubmed/21520341>`_

* `DNA-Seq Pipeline: <https://bitbucket.org/mugqic/genpipes/downloads/MUGQIC_Bioinfo_DNA-Seq.pptx>`_ some interesting information.

* `Manta - Rapid Detection of Structural Variants <https://pubmed.ncbi.nlm.nih.gov/26647377/>`_, `Using Manta for filtering and annotations <https://pmbio.org/module-05-somatic/0005/03/02/Somatic_SV_FilteringAnnotationReview/>`_.

----

.. The following are replacement texts used in this file

.. |picard_sam_to_fastq| replace:: `Picard SAM to FastQ`_
.. |sym_link_fastq| replace:: `Sym Link FastQ`_
.. |trimmomatic| replace:: `Step Trimmomatic`_
.. |merge_trimmomatic_stats| replace:: `Merge Trimmomatic Stats`_
.. |skewer_trim| replace:: `Skewer Trimming`_
.. |bwa_mem_sambamba_sort_sam| replace:: `BWA SAMbamba Sort SAM`_
.. |sambamba_merge| replace:: `SAMBAM Merge SAM Files`_
.. |gatk_indel_realigner| replace:: `GATK Indel Re-aligner`_
.. |sambamba_merge_realigned| replace:: `SAMBAM Merge Realigned`_
.. |fix_mate_by_coor| replace:: `Fix Mate by Coordinate`_
.. |picard_mark_dup| replace:: `Picard Mark Duplicates`_
.. |recalib| replace:: `Recalibration`_
.. |sym_link_final_bam| replace:: `Sym Link Final BAM`_
.. |m_dna_picard| replace:: `Metrics DNA Picard`_
.. |m_dna_sample| replace:: `Metrics DNA Sample Quality Map`_
.. |m_dna_sambamba| replace:: `Metrics DNA SAMBAM Flag Stats`_
.. |m_dna_fastqc| replace:: `Metrics DNA FastQC`_
.. |picard_calc_hs_m| replace:: `Picard Calculate HS Metrics`_
.. |gatk_call_loci| replace:: `GATK Callable Loci`_
.. |extract_com_snp| replace:: `Extract Common SNP Frequencies`_
.. |baf_plot| replace:: `BAF Plot`_
.. |gatk_hp_c| replace:: `GATK Haplotype Caller`_
.. |merge_call_i_gvcf| replace:: `Merge and call individual GVCF`_
.. |combine_gvcf| replace:: `Combine GVCF`_
.. |merge_call_c_gvcf| replace:: `Merge and call combined GVCF`_
.. |variant_recalib| replace:: `Variant Recalibrator`_
.. |hc_dec_norm| replace:: `Haplotype caller decompose and normalize`_
.. |hc_flag_map| replace:: `Haplotype caller flag mappability`_
.. |hc_snp_ann| replace:: `Haplotype caller SNP ID annotation`_
.. |hc_snp_eff| replace:: `Haplotype caller SNP Effect`_
.. |hc_dbnsfp_ann| replace:: `Haplotype caller dbNSFP annotation`_
.. |hc_gemini_ann| replace:: `Haplotype caller Gemini annotation`_
.. |hc_met_vcf_stat| replace:: `Haplotype caller metrics VCF stats`_
.. |run_multiqc| replace:: `Run MultiQC`_
.. |rawmpileup| replace:: `Raw MPileup`_
.. |rawmp_cat| replace:: `Compress Raw MPileup`_
.. |snp_n_indel_bcf| replace:: `SNP and indel BCF`_
.. |merge_filter_bcf| replace:: `Merge Filter BCF`_
.. |mp_dcom_norm| replace:: `MPileup decompose and normalize`_
.. |mp_flag_map| replace:: `MPileup Flag Mappability`_
.. |mp_snp_id_ann| replace:: `Mpileup SNP ID annotation`_
.. |mp_snp_eff| replace:: `MPileup SNP Effect`_
.. |mp_dbnsfp_ann| replace:: `MPileup dbNSFP annotation`_
.. |mp_gemini_ann| replace:: `MPileup Gemini annotation`_
.. |mp_m_vcf_stat| replace:: `MPileup Metrics VCF stats`_
.. |sambamba_mark_duplicates| replace:: `SAMBAM Mark Duplicates`_
.. |delly_call_filter| replace:: `Delly2 Call Filter`_
.. |delly_sv_annotation| replace:: `Delly2 SV Annotation`_
.. |manta_sv_calls| replace:: `Manta SV Calls`_
.. |manta_sv_annotation| replace:: `Manta SV Annotation`_
.. |lumpy_paired_sv| replace:: `Lumpy Paired SV`_
.. |lumpy_sv_annotation| replace:: `Lumpy SV Annotation`_
.. |wham_sv_call| replace:: `Wham SV Call`_
.. |wham_sv_annotation| replace:: `Wham SV Annotation`_
.. |cnvkit_batch| replace:: `CNVkit Batch`_
.. |cnvkit_sv_annotation| replace:: `CNVkit SV Annotation`_
.. |run_breakseq2| replace:: `Run BreakSeq2`_
.. |ensemble_metasv| replace:: `Ensemble MetaSV`_
.. |metasv_sv_annotation| replace:: `MetaSV Annotation`_
.. |metrics| replace:: `Metrics`_
.. |m_ngscheckmate| replace:: `Metrics NGSCheckmate`_
.. |m_verify_bam_id| replace:: `Metrics Verify BAM ID`_
.. |m_vcftools_missing_i| replace:: `Metrics VCFTools Missing Individual`_
.. |m_vcftools_depth_i| replace:: `Metrics VCFTools Depth Individual`_
.. |m_gatk_sample_fp| replace:: `Metrics GATK Sample Fingerprint`_
.. |m_gatk_cluster_fp| replace:: `Metrics GATK Cluster Fingerprint`_

.. include::  repl_cram_op.inc

.. The following are html links used in this text

.. _BWA: http://bio-bwa.sourceforge.net
.. _faster than:  https://scinapse.io/papers/2103441770
.. _MAQ: http://maq.sourceforge.net
.. _GATK: https://www.ncbi.nlm.nih.gov/pubmed/25431634 
.. _bwa-mem: https://www.ncbi.nlm.nih.gov/pubmed/19451168
.. _SAMtools: https://www.ncbi.nlm.nih.gov/pubmed/19505943
.. _SAMtools software package: http://samtools.sourceforge.net
.. _Genome in a Bottle dataset: https://www.nist.gov/programs-projects/genome-bottle
.. _dbNSFP: https://www.ncbi.nlm.nih.gov/pubmed/26555599 
.. _MultiQC: https://multiqc.info
.. _Mpileup: http://samtools.sourceforge.net/mpileup.shtml 
.. _SnpEff effects: https://pcingola.github.io/SnpEff/se_introduction/
.. _Sample DNA-Seq Report: http://gqinnovationcenter.com/services/bioinformatics/tools/dnaReport/index.html
.. _long-read: https://www.nature.com/articles/s41576-020-0236-x 
.. _Structural Variation: https://www.longdom.org/open-access/structural-variation-detection-from-next-generation-sequencing-2469-9853-S1-007.pdf
