.. _docs_gp_dnaseq:

.. spelling:: 
 
     haplotype
     Mpileup
     mpileup
     bcftools

DNA Sequencing Pipeline
========================

DNA sequencing is the process of determining the sequence of nucleotides in a section of DNA. Next-generation sequencing (NGS), also known as high-throughput sequencing, allow for sequencing of DNA and RNA much more quickly and cheaply than the previously used Sanger sequencing, and as such revolutionized the study of genomics and molecular biology. The enormous amount of short reads generated by the new DNA sequencing technologies call for the development of fast and accurate read alignment programs. 

Burrows-Wheeler Alignment tool, `BWA`_, is a new read alignment package that is based on backward search with Burrows–Wheeler Transform (BWT), to efficiently align short sequencing reads against a large reference sequence such as the human genome, allowing mismatches and gaps. BWA is ∼10–20 times `faster than`_ Mapping and Assembly with Quality, `MAQ`_, while achieving similar accuracy. In addition, BWA outputs alignment in the new standard SAM (Sequence Alignment/Map) format. Variant calling and other downstream analyses after the alignment can be achieved with the open source `SAMtools software package`_.

.. contents:: :local:

----

Introduction
------------

DNA Sequencing GenPipes pipeline has been implemented optimizing the Genome Analysis Toolkit, `GATK`_ best practices standard operating protocols. This procedure entails trimming raw reads derived from whole-genome or exome data followed by alignment to a known reference, post-alignment refinements, and variant calling. Trimmed reads are aligned to a reference by the Burrows-Wheeler Aligner (BWA), `bwa-mem`_. Refinements of mismatches near insertions and deletions (indels) and base qualities are performed using GATK indels realignment and base recalibration to improve read quality after alignment. Processed reads are marked as fragment duplicates using Picard MarkDuplicates and single-nucleotide polymorphisms and small indels are identified using either GATK haplotype callers or `SAMtools`_ mpileup. 

The `Genome in a Bottle dataset`_ was used to select steps and parameters minimizing the false-positive rate and maximizing the true-positive variants to achieve a sensitivity of 99.7%, precision of 99.1%, and F1 score of 99.4%. Finally, additional annotations are incorporated using `dbNSFP`_ and / or Gemini and QC metrics are collected at various stages and visualized using `MultiQC`_. This pipeline has two different protocols, the default protocol based on the GATK variant caller, haplotype caller, (-t mugqic) and one based on the mpileup/bcftools caller (-t mpileup).

The DNA sequencing pipeline supports two trimmers: `Trimmomatic`_ and `Skewer <https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-15-182>`_.

See :ref:`More Information on DNA Sequencing` section below for details. 

----

Version
-------

|genpipes_version|

For the latest implementation and usage details refer to DNA Sequencing implementation `README file <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/dnaseq/README.md>`_ file.

----

Usage
-----

::

  dnaseq.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
            [-o OUTPUT_DIR] [-j {pbs,batch,daemon,slurm}] [-f]
            [--no-json] [--report] [--clean]
            [-l {debug,info,warning,error,critical}] [--sanity-check]
            [--container {docker, singularity} {<CONTAINER PATH>, <CONTAINER NAME>}] 
            [-t {mugqic,mpileup,light}] [-r READSETS] [-v]

**Optional Arguments**

::

  -h                    show this help message and exit
  --help                show detailed description of pipeline and steps
  -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                        config INI-style list of files; config parameters are
                        overwritten based on files order
  -s STEPS, --steps STEPS
                        step range e.g. '1-5', '3,6,7', '2,4-8'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        output directory (default: current)
  -j {pbs,batch,daemon,slurm}, --job-scheduler {pbs,batch,daemon,slurm}
                        job scheduler type (default: slurm)
  -f, --force           force creation of jobs even if up to date (default:
                        false)
  --no-json             do not create a JSON file per analysed sample to track the
                        analysis status (default: false i.e., JSON file will be
                        created)
  --report              create 'pandoc' command to merge all job markdown
                        report files in the given step range into HTML, if
                        they exist; if --report is set, --job-scheduler,
                        --force, --clean options and job up-to-date status are
                        ignored (default: false)
  --clean               create 'rm' commands for all job removable files in
                        the given step range, if they exist; if --clean is
                        set, --job-scheduler, --force options and job up-to-
                        date status are ignored (default: false)
  -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                        log level (default: info)
  --sanity-check        run the pipeline in `sanity check mode` to verify that
                        all the input files needed for the pipeline to run are
                        available on the system (default: false)
  --container {docker, singularity} {<CONTAINER PATH>, <CONTAINER NAME>}
                        run pipeline inside a container providing a container
                        image path or accessible docker/singularity hub path
  -t {mugqic,mpileup,light}, --type {mugqic,mpileup,light}
                        DNAseq analysis type
  -r READSETS, --readsets READSETS
                        readset file
  -v, --version         show the version information and exit

----
 
Example Run
-----------

Use the following commands to execute MUGQIC DNA sequencing pipeline:

::

  dnaseq.py -t mugqic -c $MUGQIC_PIPELINES_HOME/pipelines/dnaseq/dnaseq.base.ini $MUGQIC_PIPELINES_HOME/pipelines/dnaseq/dnaseq.guillimin.ini -r readset.dnaseq.txt -s 1-29 > dnaseqCommands_mugqic.sh

  bash dnaseqCommands_mugqic.sh

Use the following commands to execute the Mpileup DNA sequencing pipeline:

::
 
  dnaseq.py -t mpileup -c $MUGQIC_PIPELINES_HOME/pipelines/dnaseq/dnaseq.base.ini $MUGQIC_PIPELINES_HOME/pipelines/dnaseq/dnaseq.guillimin.ini -r readset.dnaseq.txt -s 1-33 > dnaseqCommands_mpileup.sh

  bash dnaseqCommands_mpileup.sh

You can download the test dataset for this pipeline :ref:`here<docs_testdatasets>`.

---- 

Pipeline Schema
---------------

Figure below shows the schema of the DNA sequencing protocol - MUGQIC type. You can refer to the latest `pipeline implementation <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/dnaseq/>`_ and see here to download a high resolution image of `DNA Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_dnaseq_mugqic.png>`_ to download a high resolution image of the same.

.. figure:: /img/pipelines/dnaseq-mugqic.png
   :align: center
   :alt: dnaseq1 schema

   Figure: Schema of MUGQIC DNA Sequencing protocol

The following figure shows the pipeline schema for `Mpileup`_ type of DNA sequencing protocol. Refer to the `high resolution image in GenPipes sources <https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_dnaseq_mpileup.png>`_.

.. figure:: /img/pipelines/dnaseq-mpileup.png
   :align: center
   :alt: dnaseq2 schema 

   Figure: Schema of Mpileup DNA Sequencing protocol

There is another choice of DNA Sequencing protocol called DNA Light.  Refer to the Pipeline steps for details.

----

Pipeline Steps
--------------

The table below shows various steps that constitute the MUGQIC and Mpileup types of DNA Sequencing for genomic analysis pipelines.

+----+--------------------------------+---------------------------------+---------------------------------+
|    |  *MUGQIC DNA Sequencing Steps* |   *MPileup DNA Sequencing Steps*| *Light DNA Sequencing Steps*    |
+====+================================+=================================+=================================+
| 1. | |picard_sam_to_fastq|          | |picard_sam_to_fastq|           | |picard_sam_to_fastq|           |
+----+--------------------------------+---------------------------------+---------------------------------+
| 2. | |sym_link_fastq|               | |sym_link_fastq|                | |skewer_trim|                   |
+----+--------------------------------+---------------------------------+---------------------------------+
| 3. | |trimmomatic|                  | |trimmomatic|                   | |bwa_mem_picard_sort_sam|       |
+----+--------------------------------+---------------------------------+---------------------------------+
| 4. | |merge_trimmomatic_stats|      | |merge_trimmomatic_stats|       | |sambamba_merge|                |
+----+--------------------------------+---------------------------------+---------------------------------+
| 5. | |skewer_trim|                  | |skewer_trim|                   | |gatk_indel_realigner|          |
+----+--------------------------------+---------------------------------+---------------------------------+
| 6. | |bwa_mem_picard_sort_sam|      | |bwa_mem_picard_sort_sam|       | |sambamba_merge_realigned|      |
+----+--------------------------------+---------------------------------+---------------------------------+
| 7. | |sambamba_merge|               | |sambamba_merge|                | |sambamba_mark_duplicates|      |
+----+--------------------------------+---------------------------------+---------------------------------+
| 8. | |gatk_indel_realigner|         | |gatk_indel_realigner|          | |m_dna_picard|                  |
+----+--------------------------------+---------------------------------+---------------------------------+
| 9. | |sambamba_merge_realigned|     | |sambamba_merge_realigned|      | |m_dna_sample|                  |
+----+--------------------------------+---------------------------------+---------------------------------+
| 10.| |fix_mate_by_coor|             | |fix_mate_by_coor|              | |m_dna_sambamba|                |
+----+--------------------------------+---------------------------------+---------------------------------+
| 11.| |picard_mark_dup|              | |picard_mark_dup|               | |m_dna_fastqc|                  |
+----+--------------------------------+---------------------------------+---------------------------------+
| 12.| |recalib|                      | |recalib|                       | |picard_calc_hs_m|              |
+----+--------------------------------+---------------------------------+---------------------------------+
| 13.| |sym_link_final_bam|           | |sym_link_final_bam|            | |gatk_call_loci|                |
+----+--------------------------------+---------------------------------+---------------------------------+
| 14.| |m_dna_picard|                 | |m_dna_picard|                  | |extract_com_snp|               |
+----+--------------------------------+---------------------------------+---------------------------------+
| 15.| |m_dna_sample|                 | |m_dna_sample|                  | |baf_plot|                      |
+----+--------------------------------+---------------------------------+---------------------------------+
| 16.| |m_dna_sambamba|               | |m_dna_sambamba|                | |gatk_hp_c|                     |
+----+--------------------------------+---------------------------------+---------------------------------+
| 17.| |m_dna_fastqc|                 | |m_dna_fastqc|                  | |merge_call_i_gvcf|             |
+----+--------------------------------+---------------------------------+---------------------------------+
| 18.| |picard_calc_hs_m|             | |picard_calc_hs_m|              | |combine_gvcf|                  |
+----+--------------------------------+---------------------------------+---------------------------------+
| 19.| |gatk_call_loci|               | |gatk_call_loci|                | |merge_call_c_gvcf|             |
+----+--------------------------------+---------------------------------+---------------------------------+
| 20.| |extract_com_snp|              | |extract_com_snp|               | |combine_gvcf|                  |
+----+--------------------------------+---------------------------------+---------------------------------+
| 21.| |baf_plot|                     | |baf_plot|                      | |variant_recalib|               | 
+----+--------------------------------+---------------------------------+---------------------------------+
| 22.| |gatk_hp_c|                    | |rawmpileup|                    | |hc_dec_norm|                   |
+----+--------------------------------+---------------------------------+---------------------------------+
| 23.| |merge_call_i_gvcf|            | |rawmp_cat|                     | |hc_flag_map|                   |
+----+--------------------------------+---------------------------------+---------------------------------+
| 24.| |combine_gvcf|                 | |snp_n_indel_bcf|               | |hc_snp_ann|                    |
+----+--------------------------------+---------------------------------+---------------------------------+
| 25.| |merge_call_c_gvcf|            | |merge_filter_bcf|              | |hc_snp_eff|                    |
+----+--------------------------------+---------------------------------+---------------------------------+
| 26.| |variant_recalib|              | |mp_dcom_norm|                  | |hc_dbnsfp_ann|                 | 
+----+--------------------------------+---------------------------------+---------------------------------+
| 27.| |hc_dec_norm|                  | |mp_flag_map|                   | |hc_gemini_ann|                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 28.| |hc_flag_map|                  | |mp_snp_id_ann|                 | |run_multiqc|                   | 
+----+--------------------------------+---------------------------------+---------------------------------+
| 29.| |hc_snp_ann|                   | |mp_snp_eff|                    | |cram_output|                   |
+----+--------------------------------+---------------------------------+---------------------------------+
| 30 | |hc_snp_eff|                   | |mp_dbnsfp_ann|                 |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 31.| |hc_dbnsfp_ann|                | |mp_gemini_ann|                 |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 32.| |hc_gemini_ann|                | |mp_m_vcf_stat|                 |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 33.| |hc_met_vcf_stat|              | |run_multiqc|                   |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 34.| |run_multiqc|                  | |cram_output|                   |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+
| 35.| |cram_output|                  |                                 |                                 |
+----+--------------------------------+---------------------------------+---------------------------------+

----

.. include:: steps_dnaseq.inc

----

.. _More Information on DNA Sequencing:

More information
-----------------

For the latest implementation and usage details refer to DNA Sequencing implementation `README.md <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/dnaseq/README.md>`_ file.

* `Three-stage quality control strategies for DNA sequencing <https://academic.oup.com/bib/article/15/6/879/180439>`_

* `NGS Mapping, errors and quality control <http://bioinformatics.org.au/ws14/wp-content/uploads/ws14/sites/5/2014/07/Felicity-Newell_presentation.pdf>`_

* `dbNSFP: a lightweight database of human nonsynonymous SNPs and their functional predictions <https://www.ncbi.nlm.nih.gov/pubmed/21520341>`_

----

.. The following are replacement texts used in this file

.. |picard_sam_to_fastq| replace:: `Picard SAM to FastQ`_
.. |sym_link_fastq| replace:: `Sym Link FastQ`_
.. |trimmomatic| replace:: `Step Trimmomatic`_
.. |merge_trimmomatic_stats| replace:: `Merge Trimmomatic Stats`_
.. |skewer_trim| replace:: `Skewer Trimming`_
.. |bwa_mem_picard_sort_sam| replace:: `BWA Picard Sort SAM`_
.. |sambamba_merge| replace:: `SAMBAM Merge SAM Files`_
.. |gatk_indel_realigner| replace:: `GATK Indel Re-aligner`_
.. |sambamba_merge_realigned| replace:: `SAMBAM Merge Realigned`_
.. |fix_mate_by_coor| replace:: `Fix Mate by Coordinate`_
.. |picard_mark_dup| replace:: `Picard Mark Duplicates`_
.. |recalib| replace:: `Recalibration`_
.. |sym_link_final_bam| replace:: `Sym Link Final BAM`_
.. |m_dna_picard| replace:: `Metrics DNA Picard`_
.. |m_dna_sample| replace:: `Metrics DNA Sample Quality Map`_
.. |m_dna_sambamba| replace:: `Metrics DNA SAMBAM Flag Stats`_
.. |m_dna_fastqc| replace:: `Metrics DNA FastQC`_
.. |picard_calc_hs_m| replace:: `Picard Calculate HS Metrics`_
.. |gatk_call_loci| replace:: `GATK Callable Loci`_
.. |extract_com_snp| replace:: `Extract Common SNP Frequencies`_
.. |baf_plot| replace:: `BAF Plot`_
.. |gatk_hp_c| replace:: `GATK Haplotype Caller`_
.. |merge_call_i_gvcf| replace:: `Merge and call individual GVCF`_
.. |combine_gvcf| replace:: `Combine GVCF`_
.. |merge_call_c_gvcf| replace:: `Merge and call combined GVCF`_
.. |variant_recalib| replace:: `Variant Recalibrator`_
.. |hc_dec_norm| replace:: `Haplotype caller decompose and normalize`_
.. |hc_flag_map| replace:: `Haplotype caller flag mappability`_
.. |hc_snp_ann| replace:: `Haplotype caller SNP ID annotation`_
.. |hc_snp_eff| replace:: `Haplotype caller SNP Effect`_
.. |hc_dbnsfp_ann| replace:: `Haplotype caller dbNSFP annotation`_
.. |hc_gemini_ann| replace:: `Haplotype caller Gemini annotation`_
.. |hc_met_vcf_stat| replace:: `Haplotype caller metrics VCF stats`_
.. |run_multiqc| replace:: `Run MultiQC`_
.. |rawmpileup| replace:: `Raw MPileup`_
.. |rawmp_cat| replace:: `Compress Raw MPileup`_
.. |snp_n_indel_bcf| replace:: `SNP and indel BCF`_
.. |merge_filter_bcf| replace:: `Merge Filter BCF`_
.. |mp_dcom_norm| replace:: `MPileup decompose and normalize`_
.. |mp_flag_map| replace:: `MPileup Flag Mappability`_
.. |mp_snp_id_ann| replace:: `Mpileup SNP ID annotation`_
.. |mp_snp_eff| replace:: `MPileup SNP Effect`_
.. |mp_dbnsfp_ann| replace:: `MPileup dbNSFP annotation`_
.. |mp_gemini_ann| replace:: `MPileup Gemini annotation`_
.. |mp_m_vcf_stat| replace:: `MPileup Metrics VCF stats`_
.. |sambamba_mark_duplicates| replace:: `SAMBAM Mark Duplicates`_

.. include::  repl_cram_op.inc

.. The following are html links used in this text

.. _BWA: http://bio-bwa.sourceforge.net
.. _faster than:  https://scinapse.io/papers/2103441770
.. _MAQ: http://maq.sourceforge.net
.. _GATK: https://www.ncbi.nlm.nih.gov/pubmed/25431634 
.. _bwa-mem: https://www.ncbi.nlm.nih.gov/pubmed/19451168
.. _SAMtools: https://www.ncbi.nlm.nih.gov/pubmed/19505943
.. _SAMtools software package: http://samtools.sourceforge.net
.. _Genome in a Bottle dataset: https://www.nist.gov/programs-projects/genome-bottle
.. _dbNSFP: https://www.ncbi.nlm.nih.gov/pubmed/26555599 
.. _MultiQC: https://multiqc.info
.. _Mpileup: http://samtools.sourceforge.net/mpileup.shtml 

