.. _docs_gp_rnaseq:

.. spelling::

     html
     goseq
 
RNA Sequencing Pipeline
========================

This pipeline aligns reads with `STAR <https://www.ncbi.nlm.nih.gov/pubmed/23104886>`_ 2-passes mode, assembles transcripts with `Cufflinks <https://www.ncbi.nlm.nih.gov/pubmed/20436464>`_, and performs differential expression with `Cuffdiff <https://www.ncbi.nlm.nih.gov/pubmed/23222703>`_. In parallel, gene-level expression is quantified using `htseq-count <https://www.ncbi.nlm.nih.gov/pubmed/25260700>`_, which produces raw read counts that are subsequently used for differential gene expression with both `DESeq process <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3218662/>`_ and `edgeR algorithms <https://www.ncbi.nlm.nih.gov/pubmed/19910308>`_. Several common quality metrics (e.g., ribosomal RNA content, expression saturation estimation) are also calculated through the use of `RNA-SeQC <https://www.ncbi.nlm.nih.gov/pubmed/22539670>`_ and in-house scripts. Gene Ontology terms are also tested for over-representation using `GOseq <https://www.ncbi.nlm.nih.gov/pubmed/20132535>`_. Expressed short single-nucleotide variants (SNVs) and indels calling is also performed by this pipeline, which optimizes GATK best practices to reach a sensitivity of 92.8%, precision of 87.7%, and F1 score of 90.1%. 

.. contents:: :local:

----

Introduction
------------

The standard MUGQIC RNA-Seq pipeline is based on the use of the `STAR aligner <https://code.google.com/p/rna-star/>`_ to align reads to the reference genome. These alignments are used during downstream analysis to determine genes and transcripts differential expression. The `Cufflinks suite <http://cufflinks.cbcb.umd.edu/>`_ is used for the transcript analysis whereas `DESeq <http://bioconductor.org/packages/release/bioc/html/DESeq.html>`_ and `edgeR <http://bioconductor.org/packages/release/bioc/html/edgeR.html>`_ are used for the gene analysis.  The RNAseq pipeline requires to provide a :ref:`design<docs_design_file>` file which will be used to define group comparison in the differential analyses.  The differential gene analysis is followed by a Gene Ontology (GO) enrichment analysis. This analysis use the `goseq approach <http://bioconductor.org/packages/release/bioc/html/goseq.html>`_. The goseq is based on the use of non-native GO terms (see details in the section 5 of the `corresponding vignette <http://bioconductor.org/packages/release/bioc/vignettes/goseq/inst/doc/goseq.pdf>`_.

Finally, a summary html report is automatically generated by the pipeline at the end of the analysis. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software tools and methods used during the analysis, together with the full list of parameters that have been passed to the pipeline main script.

See :ref:`More Information` section below for details.

----

Version
-------

|genpipes_version| 

For the latest implementation and usage details refer to RNA Sequencing implementation `README file <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/rnaseq/README.md>`_ file.

----

Usage
-----

::

  usage: rnaseq.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
                 [-o OUTPUT_DIR] [-j {pbs,batch,daemon,slurm}] [-f] [--json]
                 [--report] [--clean] [-l {debug,info,warning,error,critical}]
                 [-d DESIGN] [-t {cufflinks,stringtie}] [-r READSETS] [-v]

**Optional Arguments**

::

    -h                    show this help message and exit
    --help                show detailed description of pipeline and steps
    -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                          config INI-style list of files; config parameters are
                          overwritten based on files order
    -s STEPS, --steps STEPS
                          step range e.g. '1-5', '3,6,7', '2,4-8'
    -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                          output directory (default: current)
    -j {pbs,batch,daemon,slurm}, --job-scheduler {pbs,batch,daemon,slurm}
                          job scheduler type (default: slurm)
    -f, --force           force creation of jobs even if up to date (default:
                          false)
    --json                create a JSON file per analysed sample to track the
                          analysis status (default: false)
    --report              create 'pandoc' command to merge all job markdown
                          report files in the given step range into HTML, if
                          they exist; if --report is set, --job-scheduler,
                          --force, --clean options and job up-to-date status are
                          ignored (default: false)
    --clean               create 'rm' commands for all job removable files in
                          the given step range, if they exist; if --clean is
                          set, --job-scheduler, --force options and job up-to-
                          date status are ignored (default: false)
    -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                          log level (default: info)
    -d DESIGN, --design DESIGN
                          design file
    -t {cufflinks,stringtie}, --type {cufflinks,stringtie}
                          Type of RNA-seq assembly method (default cufflinks)
    -r READSETS, --readsets READSETS
                          readset file
    -v, --version         show the version information and exit

----

Example Run
-----------

You can download `RNA Sequencing Pipeline test dataset <http://www.computationalgenomics.ca/tutorial/c3g_analysis_workshop/C3GAW_RNA_TestData_Aug2018.zip>`_ and use the following command to execute the RNA Sequencing genomics pipeline:

::

  rnaseq.py -c $MUGQIC_PIPELINES_HOME/pipelines/rnaseq/rnaseq.base.ini 
               $MUGQIC_PIPELINES_HOME/pipelines/rnaseq/rnaseq.cedar.ini workshop.ini
               -r readset.rnaseq.txt -d design.rnaseq.txt -s 1-24 -j slurm > commands.txt

  bash commands.txt 

This set of commands is meant for running GenPipes on C3 data center. For more details, you can refer to GenPipes `RNA Sequencing Workshop 2018 presentation <http://www.computationalgenomics.ca/tutorial/c3g_analysis_workshop/C3GAW_RNASeq_Aug2018.zip>`_.

----

Pipeline Schema
---------------

Figure below shows the schema of the RNA sequencing protocol. You can refer to the latest `pipeline implementation <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/rnaseq/>`_ and see here to download a high resolution image of `RNA Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_rnaseq.png>`_ to download a high resolution image of the same.

.. figure:: /img/pipelines/rnaseq.png
   :align: center
   :alt: rnaseq schema

   Figure: Schema of RNA Sequencing pipeline

----

Pipeline Steps
--------------

The table below lists various steps of GenPipes RNA Sequencing Pipeline:

+----+-------------------------------------------+----------------------------------------+
|    |    *RNA Sequencing (Cufflinks)*           |   *RNA Sequencing (Stringtie)*         |
+====+===========================================+========================================+
| 1. | |picard_sam_to_fastq|                     |  |picard_sam_to_fastq|                 |
+----+-------------------------------------------+----------------------------------------+
| 2. | |trimmomatic|                             |  |trimmomatic|                         |
+----+-------------------------------------------+----------------------------------------+
| 3. | |merge_trimmomatic_stats|                 |  |merge_trimmomatic_stats|             |
+----+-------------------------------------------+----------------------------------------+
| 4. | |star_step|                               |  |star_step|                           |
+----+-------------------------------------------+----------------------------------------+
| 5. | |picard_merge_sam_files|                  |  |picard_merge_sam_files|              |
+----+-------------------------------------------+----------------------------------------+
| 6. | |picard_sort_sam|                         |  |picard_sort_sam|                     |
+----+-------------------------------------------+----------------------------------------+
| 7. | |picard_mark_duplicates|                  |  |picard_mark_duplicates|              |
+----+-------------------------------------------+----------------------------------------+
| 8. | |picard_rna_metrics|                      |  |picard_rna_metrics|                  |
+----+-------------------------------------------+----------------------------------------+
| 9. | |estimate_ribosomal_rna|                  |  |estimate_ribosomal_rna|              |
+----+-------------------------------------------+----------------------------------------+
| 10.| |bam_hard_clip|                           |  |bam_hard_clip|                       |
+----+-------------------------------------------+----------------------------------------+
| 11.| |rnaseqc|                                 |  |rnaseqc|                             |
+----+-------------------------------------------+----------------------------------------+
| 12.| |wiggle|                                  |  |wiggle|                              |
+----+-------------------------------------------+----------------------------------------+
| 13.| |raw_counts|                              |  |raw_counts|                          |
+----+-------------------------------------------+----------------------------------------+
| 14.| |raw_counts_metrics|                      |  |raw_counts_metrics|                  |
+----+-------------------------------------------+----------------------------------------+
| 15.| |cufflinks|                               |  |stringtie|                           |
+----+-------------------------------------------+----------------------------------------+
| 16.| |cuffmerge|                               |  |stringtie_merge|                     |
+----+-------------------------------------------+----------------------------------------+
| 17.| |cuffquant|                               |  |stringtie_abund|                     |
+----+-------------------------------------------+----------------------------------------+
| 18.| |cuffdiff|                                |  |ballgown|                            |
+----+-------------------------------------------+----------------------------------------+
| 19.| |cuffnorm|                                |  |differential_expression|             |
+----+-------------------------------------------+----------------------------------------+
| 20.| |fpkm_correlation_matrix|                 |                                        |
+----+-------------------------------------------+                                        |
| 21.| |gq_seq_utils_exploratory_analysis_rnaseq||                                        |
+----+-------------------------------------------+                                        |
| 22.| |differential_expression|                 |                                        |
+----+-------------------------------------------+                                        |
| 23.| |differential_expression_goseq|           |                                        |
+----+-------------------------------------------+                                        |
| 24.| |ihec_metrics|                            |                                        |
+----+-------------------------------------------+                                        |
| 25.| |verify_bam_id|                           |                                        |
+----+-------------------------------------------+----------------------------------------+

----

.. include:: steps_rnaseq.inc

----

.. _More Information on RNA Sequencing:

More information
-----------------

For the latest implementation and usage details refer to `README.md <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/rnaseq/README.md>`_.

* `RNA Sequencing Workshop 2019 (Slides) <http://www.computationalgenomics.ca/analysis-workshop-2019-rnaseq-module/>`_

----

.. The following are replacement texts used in this file

.. |picard_sam_to_fastq| replace:: `Picard SAM to FastQ`_
.. |trimmomatic| replace:: `Trimmomatic`_
.. |merge_trimmomatic_stats| replace:: `Merge Trimmomatic Stats`_
.. |star_step| replace:: `Star Processing`_
.. |picard_merge_sam_files| replace:: `Picard Merge SAM Files`_
.. |picard_sort_sam| replace:: `Picard Sort SAM`_
.. |picard_mark_duplicates| replace:: `Picard Mark Duplicates`_
.. |picard_rna_metrics| replace:: `Picard RNA Metrics`_
.. |estimate_ribosomal_rna| replace:: `Estimate Ribosomal RNA`_
.. |bam_hard_clip| replace:: `BAM Hard Clip`_
.. |rnaseqc| replace:: `RNA Seq Compress`_
.. |wiggle| replace:: `Wiggle`_
.. |raw_counts| replace:: `Raw Counts`_
.. |raw_counts_metrics| replace:: `Raw Counts Metrics`_
.. |cufflinks| replace:: `Cufflinks Process`_
.. |cuffmerge| replace:: `Cuffmerge Process`_
.. |cuffquant| replace:: `Cuffquant Step`_
.. |cuffdiff| replace:: `Cuffdiff Process`_
.. |cuffnorm| replace:: `Cuffnorm Normalization`_
.. |fpkm_correlation_matrix| replace:: `FPKM Correlation`_
.. |gq_seq_utils_exploratory_analysis_rnaseq| replace:: `GQ RNA Sequencing Utility`_
.. |differential_expression| replace:: `Differential Expression`_
.. |differential_expression_goseq| replace:: `Differential Expression GO sequencing`_
.. |ihec_metrics| replace:: `IHEC Metrics`_
.. |verify_bam_id| replace:: `Verify BAM ID`_
.. |stringtie| replace:: `Stringtie`_
.. |stringtie_abund| replace:: `Stringtie Assemble Transcriptome`_
.. |stringtie_merge| replace:: `Stringtie Merge`_
.. |ballgown| replace:: `Ballgown Gene Expression`_
