.. _docs_methylation:

.. spelling::

     epigenetic
     methylome
     Cytosine
     cytosines
     CpG
     mappability 

Methylation Sequencing Pipeline
===============================

`DNA Methylation`_ is an important epigenetic system related to gene expression. It is involved in embryonic development, genomic imprinting, X chromosome inactivation and cell differentiation. Since methylation takes part in many normal cellular functions, aberrant methylation of DNA may be associated with a wide spectrum of human diseases, including cancer. Bisulfite sequencing approaches are currently considered a “gold standard” allowing comprehensive understanding of the methylome landscape. Whole-genome bisulfite sequencing (WGBS), provides single-base resolution of methylated cytosines across the entire genome.

GenPipes methylation sequencing workflow is adapted from the `Bismark pipeline`_. It aligns paired-end reads with `Bowtie 2`_ default mode. Duplicates are removed with `Picard`_, and methylation calls are extracted using Bismark. `Wiggle`_ tracks for both read coverage and methylation profile are generated for visualization. Variant calls can be extracted from the whole-genome bisulfite sequencing (WGBS) data directly using `Bis-SNP caller`_. Bisulfite conversion rates are estimated with lambda genome or from human non-CpG methylation directly. Several metrics based on IHEC requirements are also calculated. Methylation sequencing can also process capture data if provided with a capture `BED file`_. 

Both paired-end reads as well as single-end reads are supported by this pipeline. Paired-end reading improves the ability to identify the relative positions of various reads in the genome making it much more effective than single-end reading for resolving structural rearrangements such as gene insertions, deletions or inversions and assembly of repetitive regions.  Single-end reads are much less expensive in terms of resource consumption and time needed for analysis and can be used for experiments that do not require higher degrees of accuracy offered by paired-reads.

.. contents:: :local:

----

Introduction
------------

The standard MUGQIC Methyl-Seq pipeline uses Bismark to align reads to the reference genome. Treatment and filtering of mapped reads approaches as mark duplicate reads, recalibration and sort are executed using `Picard`_ and GATK. Samtools MPILEUP and `BCFtool`_ are used to produce the standard SNP and indels variants file (VCF). Additional SVN annotations mostly applicable to human samples include mappability flags, `dbSNP annotation`_ and extra information about SVN by using published databases. The `SNPeff tool`_ is used to annotate variants using an integrated database of functional predictions from multiple algorithms (`SIFT`_, `Polyphen 2`_, `LRT`_ and `MutationTaster`_, `PhyloP`_ and `GERP++`, etc.) and to calculate the effects they produce on known genes.

A summary html report is automatically generated by the pipeline. This report contains description of the sequencing experiment as well as a detailed presentation of the pipeline steps and results. Various Quality Control (QC) summary statistics are included in the report and additional QC analysis is accessible for download directly through the report. The report includes also the main references of the software and methods used during the analysis, together with the full list of parameters that have been passed to the pipeline main script.

----

Version
-------

|genpipes_version|

For the latest implementation and usage details refer to WGS or Methylation Sequencing implementation `README file <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/methylseq/README.md>`_ file.

----

Usage
-----

::

  methylseq.py [-h] [--help] [-c CONFIG [CONFIG ...]] [-s STEPS]
                    [-o OUTPUT_DIR] [-j {pbs,batch,daemon,slurm}] [-f]
                    [--no-json] [--report] [--clean]
                    [-l {debug,info,warning,error,critical}] [--sanity-check]
                    [--container {docker, singularity} {<CONTAINER PATH>, <CONTAINER NAME>}] 
                    [-d DESIGN]
                    [-t {mugqic,mpileup,light}] [-r READSETS] [-v]


**Optional Arguments**

::

  -h                    show this help message and exit
  --help                show detailed description of pipeline and steps
  -c CONFIG [CONFIG ...], --config CONFIG [CONFIG ...]
                        config INI-style list of files; config parameters are
                        overwritten based on files order
  -s STEPS, --steps STEPS
                        step range e.g. '1-5', '3,6,7', '2,4-8'
  -o OUTPUT_DIR, --output-dir OUTPUT_DIR
                        output directory (default: current)
  -j {pbs,batch,daemon,slurm}, --job-scheduler {pbs,batch,daemon,slurm}
                        job scheduler type (default: slurm)
  -f, --force           force creation of jobs even if up to date (default:
                        false)
  --no-json             do not create a JSON file per analysed sample to track the
                        analysis status (default: false i.e. JSON file will be created)
  --report              create 'pandoc' command to merge all job markdown
                        report files in the given step range into HTML, if
                        they exist; if --report is set, --job-scheduler,
                        --force, --clean options and job up-to-date status are
                        ignored (default: false)
  --clean               create 'rm' commands for all job removable files in
                        the given step range, if they exist; if --clean is
                        set, --job-scheduler, --force options and job up-to-
                        date status are ignored (default: false)
  -l {debug,info,warning,error,critical}, --log {debug,info,warning,error,critical}
                        log level (default: info)
  --sanity-check        run the pipeline in `sanity check mode` to verify that 
                        all the input files needed for hte pipeline to run are
                        available on the system (default: false)
  --container {docker, singularity,} {<CONTAINER PATH>, <CONTAINER NAME>}
                        run pipeline inside a container providing a container
                        image path or accessible docker/singularity hub path
  -d DESIGN, --design DESIGN
                        design file
  -t {mugqic,mpileup,light}, --type {mugqic,mpileup,light}
                        DNAseq analysis type
  -r READSETS, --readsets READSETS
                        readset file
  -v, --version         show the version information and exit

----

Example Run
-----------

Use the following commands to execute the methylation pipeline:

::

  methylseq.py -c $MUGQIC_PIPELINES_HOME/pipelines/methylseq/methylseq.base.ini $MUGQIC_PIPELINES_HOME/pipelines/methylseq/methylseq.guillimin.ini -r readset.methylseq.txt -s 1-14 > methylseq.sh

  bash methylseq.sh

You can download the test dataset for this pipeline :ref:`in Reference section<docs_testdatasets>`.

Pipeline Schema
---------------
Figure below shows the schema of the WGS or Methylation sequencing pipeline. You can refer to the latest `pipeline implementation <https://bitbucket.org/mugqic/genpipes/raw/master/resources/workflows/GenPipes_methylseq.png>`_ to download a high resolution image of the same.

.. figure:: /img/pipelines/wgs_methylseq.png
   :align: center
   :alt: methylation_seq 

   Figure: Schema of WGS or Methylation Sequencing pipeline


Pipeline Steps
--------------

The table below shows various steps that constitute the WGS or Methylation Sequencing genomic analysis pipelines.

+----+--------------------------------------+
|    |  *Methylation sequencing Steps*      |
+====+======================================+
| 1. | |picard_sam_to_fastq|                |
+----+--------------------------------------+
| 2. | |trimmomatic|                        |
+----+--------------------------------------+
| 3. | |merge_trimmomatic_stats|            |
+----+--------------------------------------+
| 4. | |bismark_align|                      |
+----+--------------------------------------+
| 5. | |add_bam_umi|                        |
+----+--------------------------------------+
| 6. | |picard_merge_sam_files|             |
+----+--------------------------------------+
| 7. | |picard_remove_duplicates|           |
+----+--------------------------------------+
| 8. | |metrics|                            |
+----+--------------------------------------+
| 9. | |methylation_call|                   |
+----+--------------------------------------+
| 10.| |wiggle_tracks|                      |
+----+--------------------------------------+
| 11.| |methylation_profile|                |
+----+--------------------------------------+
| 12.| |ihec_sample_metrics_report|         |
+----+--------------------------------------+
| 13.| |bis_snp|                            |
+----+--------------------------------------+
| 14.| |filter_snp_cpg|                     |
+----+--------------------------------------+
| 15.| |prepare_methylkit|                  |
+----+--------------------------------------+
| 16.| |methylkit_differential_analysis|    |
+----+--------------------------------------+
| 16.| |cram_output|                        |
+----+--------------------------------------+


----

.. include:: steps_methylseq.inc

----

.. _More Information Methylseq:

More information
-----------------

For the latest implementation and usage details refer to WGS or Methylation Sequencing implementation `README.md <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/methylseq/README.md>`_ file.

* `DNA Methylation Detection: Bisulphite genomic sequencing analysis <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3233226/>`_.

* `Strategies for analyzing bisulfite sequencing data <https://www.sciencedirect.com/science/article/pii/S0168165617315936>`_.

* `Analysis and Visualization Tool for Targeted Amplicon Bisulfite Sequencing on Ion Torrent Sequencers <https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0160227>`_.

* Cytosine and Guanine - `CpGs`_.

.. The following are html links used in this text

.. _DNA Methylation: https://genestack-user-tutorials.readthedocs.io/tutorials/Methylation_profiling/
.. _Bismark pipeline: https://www.ncbi.nlm.nih.gov/pubmed/21493656
.. _Bis-SNP caller: https://www.ncbi.nlm.nih.gov/pubmed/22784381
.. _BED file: https://asia.ensembl.org/info/website/upload/bed.html
.. _CpGs: https://en.wikipedia.org/wiki/CpG_site
.. _Bowtie 2: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml
.. _BCFtool: https://samtools.github.io/bcftools/bcftools.html
.. _dbSNP annotation: https://www.ncbi.nlm.nih.gov/books/NBK44455/
.. _Polyphen 2: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4480630/
.. _PhyloP: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2798823/
.. _Picard: https://www.broadinstitute.org/files/shared/mpg/plathumgen/plathumgen_fennell.pdf
.. _Wiggle: https://genome.ucsc.edu/goldenPath/help/wiggle.html
.. _SNPeff tool: http://snpeff.sourceforge.net/SnpEff.html
.. _SIFT: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC168916/
.. _LRT: https://en.wikipedia.org/wiki/Likelihood-ratio_test
.. _MutationTaster: https://en.wikipedia.org/wiki/MutationTaster
.. _GERP++: http://mendel.stanford.edu/sidowlab/downloads/gerp/index.html

.. The following are replacement texts used in this file

.. |picard_sam_to_fastq| replace:: `Picard SAM to FastQ`_
.. |trimmomatic| replace:: `Trimmomatic`_
.. |merge_trimmomatic_stats| replace:: `Merge Trimmomatic Statistics`_
.. |bismark_align| replace:: `Bismark Aligner Step`_
.. |add_bam_umi| replace:: `UMI BAM Tag Processing`_
.. |picard_merge_sam_files| replace:: `Picard merge SAM files`_
.. |picard_remove_duplicates| replace:: `Picard remove duplicates`_
.. |metrics| replace:: `Compute Metrics`_
.. |methylation_call| replace:: `Methylation Call`_
.. |wiggle_tracks| replace:: `Wiggle Tracks`_
.. |methylation_profile| replace:: `Methylation Profile`_
.. |ihec_sample_metrics_report| replace:: `IHEC Sample Metrics Report`_
.. |bis_snp| replace:: `Bis SNP Processing`_
.. |filter_snp_cpg| replace:: `Filter SNP CpGs`_
.. |prepare_methylkit| replace:: `Prepare for MethylKit Differential Analysis`_
.. |methylkit_differential_analysis| replace:: `MethylKit Differential Analysis`_
.. include:: repl_cram_op.inc
