:orphan:

.. _docs_gp_relnote_4_5_0:

.. spelling::

     sed
     sambamba
     genpipes
     gvcf
     bfx
     symlink
     sortmeRNA
     htseq
     deliverables
     md5sum
     md

GenPipes 4.5.0 Release Notes
============================

What's new? 
-----------

This is a medium release with bug-fixes:

**General**

#. A bug in the clean command (GenPipes v4.4.5) caused the removal of the output directory. It is now fixed. We encourage everyone to use GenPipes v4.5.0 to avoid this costly problem of losing the output directory!
#. The automatically generated `job_list` file name now contains both the pipeline and the protocol name. This makes it much easier to identify the appropriate `job_list` file when running multiple protocols. 

**Pipeline Enhancements**

* `Tumor Pair Pipeline <https://bitbucket.org/mugqic/genpipes/src/4.5.0/pipelines/tumor_pair/>`_:
  - Added `CollectWgsMetrics` step
  - Improved display of QualiMap output in `MultiQC` report 
       
* `RNA Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/src/4.5.0/pipelines/rnaseq/>`_:
  - Added `md5sum`` for the deliverables  
  - Updated `gtex` version

* `Chip Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/src/4.5.0/pipelines/chipseq/>`_:
  - A typo in error message was fixed
      
Quick! Where can I find it? I can't wait! 
------------------------------------------
 
* The latest release for GenPipes version 4.5.0 tarball is in: https://bitbucket.org/mugqic/genpipes/downloads/genpipes-4.5.0.tar.gz

* The module mugqic/genpipes/4.5.0 is installed on Abacus (MGC), Béluga, Narval, Graham and Cedar Compute Canada clusters.

* The source code is in bitbucket: https://bitbucket.org/mugqic/genpipes

* For latest GenPipes documentation, please refer to: https://genpipes.readthedocs.io/

Where is the detailed ChangeLog? 
================================= 

The `GenPipes ChangeLog <https://bitbucket.org/mugqic/genpipes/src/master/CHANGELOG.md>`_ is available in the archive as well as in the repository.

Since we use git, there are many ways to get the details in many formats. 
One of our preferred ways is to use a script written by the author of the Ray assembler: Sébastien Boisvert, 
which lists the commits by tag and author: 

https://raw.github.com/sebhtml/ray/master/scripts/dump-ChangeLog.sh 

Enjoy our pipelines on the many clusters where they are installed!
