:orphan:

.. _docs_gp_relnote_4_4_5:

.. spelling::

     sed
     sambamba
     genpipes
     gvcf
     bfx

GenPipes 4.4.5 Release Notes
============================

What's new? 
-----------

This is a bug-fix minor release:

* `Tumor Pair Pipeline <https://bitbucket.org/mugqic/genpipes/src/4.4.5/pipelines/tumor_pair/>`_:
  -  Improved dependencies and symlink generation in `sambamba_merge` step
       
* `RNA Sequencing Pipeline (Light) <https://bitbucket.org/mugqic/genpipes/src/4.4.5/pipelines/rnaseq_light/>`_:
  - Fixed an issue when using single-end reads: bootstraps were not being generated, causing the sleuth step to fail
  - Improvements in kallisto speed by adding threads option (can be configured via ini)

* `RNA Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/src/4.4.5/pipelines/rnaseq/>`_:
  - Addition of sortmeRNA metrics
  - Fixed an issue with using single-end reads: htseq-count metrics were generated with 'reverse' setting, causing incorrect read counts
  - **htseq-count** 'stranded' option can now be set via the config file
  - Updated htseq version

* `Chip Sequencing Pipeline <https://bitbucket.org/mugqic/genpipes/src/master/pipelines/chipseq/>`_:
  - In the differential binding step, thresholds for tables and figures can now be configured independently

* **Improved Throughput**
  - Fixed a bug in the ordering of modules that caused unnecessary restarting of steps
      
Quick! Where can I find it? I can't wait! 
------------------------------------------
 
* The latest release for GenPipes version 4.4.5 tarball is in: https://bitbucket.org/mugqic/genpipes/downloads/genpipes-4.4.5.tar.gz

* The module mugqic/genpipes/4.4.5 is installed on Abacus (MGC), Béluga, Narval, Graham and Cedar Compute Canada clusters.

* The source code is in bitbucket: https://bitbucket.org/mugqic/genpipes

* For latest GenPipes documentation, please refer to: https://genpipes.readthedocs.io/

Where is the detailed ChangeLog? 
================================= 

A ChangeLog (`CHANGELOG.md <https://bitbucket.org/mugqic/genpipes/src/master/CHANGELOG.md>`_) is available in the archive as well as in the repository.

Since we use git, there are many ways to get the details in many formats. 
One of our preferred ways is to use a script written by the author of the Ray assembler: Sébastien Boisvert, 
which lists the commits by tag and author: 

https://raw.github.com/sebhtml/ray/master/scripts/dump-ChangeLog.sh 

Enjoy our pipelines on the many clusters where they are installed!
