:orphan:

.. _docs_gp_relnote_6_0_0:

.. spelling:word-list::

     genpipes


GenPipes 6.0.0 Release Notes
============================

What's new? 
-----------

This is a major release of GenPipes: 

* GenPipes Source tree
   - Source tree has migrated from `Bitbucket <https://bitbucket.org/mugqic/genpipes/src/master/genpipes/>`_ to `GitHub <https://github.com/c3g/GenPipes>`_.
* :ref:`ChIPSeq<docs_gp_chipseq>`
   - The Chipseq pipeline `atac` protocol has a new step, Deeptools QC, that generates two new metrics: Fingerplot QC and Correlation Matrix. These new metrics are now included in the MultiQC report.
* A new :ref:`LongRead DNASeq pipeline<docs_gp_longread_dnaseq>`, is now available to analyze long-read sequences.
   - Supports two protocols, Revio (PacBio Revio) and Nanopore (ONT Nanopore)
   - Nanopore protocol of long read dnaseq is equivalent in functionality to the Nanopore pipeline that was available as a standalone pipeline in the previous versions
* Nanopore pipeline is deprecated as a standalone GenPipes pipeline effective v6.0. The same functionality is now available as the Nanopore protocol of the new :ref:`LongRead DNASeq pipeline<docs_gp_longread_dnaseq>`.
* :ref:`RNASeq<docs_gp_rnaseq>` minor updates include error fixes for errors that occurred when running the rnaseq_cancer protocol with single end reads.
* Output file Standardization
   - The 'Prologues' and 'Epilogues' found in the GenPipes output files have been revamped and standardized. 
   - This makes them exactly the same for SLURM and PBS clusters simplifying GenPipes usability.
* Input file validation check
   - A new tool is available to check the format of your readset and design files
   - The GenPipes tool `validate_genpipes` checks input files for common issues. It detects missing columns, formatting errors, trailing whitespaces, and more. This helps GenPipes users save time. Validation ensures runs fail quickly if issues are found, avoiding delays after the run starts.
* :ref:`Log Report Enhancements<docs_gp_job_results>`
    - It is now simpler to generate log reports for GenPipes runs, regardless of whether you're working on PBS or SLURM clusters. 
    - Instead of two different commands there is a single command to generate logs irrespective of the scheduler. A single command can be used on all clusters to generate a report of the status of each job in the GenPipes pipeline:
  
   ::
    
    genpipes tools log_report --tsv log.out job_output/{PIPELINE}_job_list_{DATE}T{TIME}

Quick! Where can I find it? I can't wait! 
------------------------------------------
 
* The latest release for GenPipes version 6.0.0 tarball is in: https://bitbucket.org/mugqic/genpipes/downloads/genpipes-6.0.0.tar.gz

* GenPipes is now distributed as a package via `pypi`: [https://pypi.org/project/c3g-genpipes/](https://pypi.org/project/c3g-genpipes/)

* The module genpipes/6.0.0 is installed on the Abacus (MGC), BÃ©luga, Graham, Cedar and Narval Alliance clusters.

* The source code is in GitHub: hhttps://github.com/c3g/GenPipes

* For latest GenPipes documentation, please refer to: https://genpipes.readthedocs.io/

Where is the detailed ChangeLog? 
================================= 

.. include:: ../gp_changelog.inc